crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs

```diff
@@ -362,111 +348,67 @@ impl RetriableRpcClient {
-                        .as_ref()
-                        .expect("fallback client must set")
-                        .skip_rpc_for_checkpoint_duration(),
-                Ordering::Relaxed,
-            );
-        }
-
-        let fallback_start_time = Instant::now();
-        let result = self
-            .call_get_full_checkpoint_from_fallback_with_retries(fallback_client, sequence_number)
-            .await;
-
-        if let Some(metrics) = self.metrics.as_ref() {
-            metrics.record_fallback_metrics(
-                "get_full_checkpoint",
-                &result,
-                fallback_start_time.elapsed(),
-            )
+        match self.get_full_checkpoint_from_primary(sequence_number).await {
+            Ok(checkpoint) => {
+                self.download_handler
+                    .record_primary_success(sequence_number);
+                Ok(checkpoint)
+            }
+            Err(primary_error) => {
+                match self
+                    .download_handler
+                    .handle_primary_checkpoint_failure(sequence_number, primary_error)
+                {
+                    CheckpointDownloadDecision::TryFallback(Some(_)) => {
+                        self.handle_fallback_attempt(sequence_number).await
+                    }
+                    CheckpointDownloadDecision::UsePrimaryError(err) => Err(err),
+                    _ => unreachable!("Invalid decision after primary failure assessment"),
+                }
+            }
         }
-
-        result
     }
 
-    /// Calls the fallback service (checkpoint archival) with retries to fetch the full checkpoint.
-    async fn call_get_full_checkpoint_from_fallback_with_retries(
+    /// Handles the fallback attempt
+    async fn handle_fallback_attempt(
         &self,
-        fallback: &FallbackClient,
         sequence_number: u64,
     ) -> Result<CheckpointData, RetriableClientError> {
-        tracing::debug!(sequence_number, "fetching checkpoint from fallback client");
-        return retry_rpc_errors(
-            self.get_strategy(),
-            || async {
-                fallback
-                    .get_full_checkpoint(sequence_number)
-                    .await
-                    .map_err(RetriableClientError::from)
-            },
-            self.metrics.clone(),
-            "get_full_checkpoint_from_fallback_with_retries",
-        )
-        .await;
+        let Some(fb_client) = &self.fallback_client else {
+            return Err(RetriableClientError::Other(anyhow::anyhow!(
+                "Misconfiguration: Skip primary indicated but no fallback client."
+            )));
+        };
+
+        let result = fb_client
+            .get_full_checkpoint(sequence_number)
+            .await
+            .map_err(RetriableClientError::from);
+
+        if result.is_ok() {
+            self.download_handler
+                .record_fallback_success_for_sequence(sequence_number);
+        }
+        result
```

Comment:
```suggestion
        Ok(fb_client
            .get_full_checkpoint(sequence_number)
            .await
            .inspect(|_| {
                self.download_handler
                    .record_fallback_success_for_sequence(sequence_number);
            })?)
```
nit: just reflowing this code a bit to avoid extra control flow logic